test/integration/filters/add_body_filter.cc

```diff
@@ -24,11 +24,21 @@ class AddBodyStreamFilter : public Http::PassThroughFilter {
       Buffer::OwnedImpl body("body");
       headers.setContentLength(body.length());
       decoder_callbacks_->addDecodedData(body, false);
+    } else {
+      headers.removeContentLength();
     }
 
     return Http::FilterHeadersStatus::Continue;
   }
 
+  Http::FilterDataStatus encodeData(Buffer::Instance& data, bool end_stream) override {
+    // For HTTP/3, the only time the protocol is set at the connection level, there's no
+    // headers-only streams so the data will be added here.
+    ASSERT(end_stream == false || decoder_callbacks_->connection()->streamInfo().protocol());
```

Comment:
This still seems a bit confusing to me, when do we expect end_stream=false and when do we expect the protocol to be set?